sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/FluxAutoComplete.java

```diff
@@ -0,0 +1,130 @@
+        private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon;
+        private final ClientLogger logger;
+
+        AutoCompleteSubscriber(CoreSubscriber<? super T> downstream,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon, ClientLogger logger) {
+            this.downstream = downstream;
+            this.onComplete = onComplete;
+            this.onAbandon = onAbandon;
+            this.logger = logger;
+        }
+
+        @Override
+        protected void hookOnSubscribe(Subscription subscription) {
+            logger.info("Subscription received. Subscribing downstream. {}", subscription);
+            downstream.onSubscribe(this);
+            requestUnbounded();
+        }
+
+        @Override
+        protected void hookOnNext(T value) {
+            final ServiceBusReceivedMessage message = value.getMessage();
+            final String sequenceNumber = message != null ? String.valueOf(message.getSequenceNumber()) : "n/a";
+
+            logger.verbose("Passing message downstream. sequenceNumber[{}]", sequenceNumber);
+
+            try {
+                downstream.onNext(value);
+                applyWithCatch(onComplete, value, "complete");
+            } catch (Exception e) {
+                logger.error("Error occurred processing message. Abandoning. sequenceNumber[{}]",
+                    sequenceNumber, e);
+
+                applyWithCatch(onAbandon, value, "abandon");
+            }
+        }
+
+        /**
+         * On an error, will pass the exception downstream.
+         *
+         * @param throwable Error to pass downstream.
+         */
+        @Override
+        protected void hookOnError(Throwable throwable) {
+            logger.error("Error occurred. Passing downstream.", throwable);
+            downstream.onError(throwable);
+        }
+
+        /**
+         * On the completion. Will pass the complete signal downstream.
+         */
+        @Override
+        protected void hookOnComplete() {
+            logger.info("Completed. Passing downstream.");
+            downstream.onComplete();
+        }
+
+        @Override
+        public Context currentContext() {
+            return downstream.currentContext();
+        }
+
+        /**
+         * Applies a function and catches then logs and closes any exceptions.
+         *
+         * @param function Function to apply.
+         * @param message received message to apply function to.
+         * @param operation The operation name.
+         */
+        private void applyWithCatch(Function<ServiceBusReceivedMessageContext, Mono<Void>> function, T message,
+            String operation) {
+            try {
+                function.apply(message).block();
+            } catch (Exception e) {
+                logger.warning("Unable to '{}' message.", operation, e);
+                onError(e);
```

Comment:
Don't think we should call `onError()` here if the operation was to `abandon`. This will terminate the downstream. What is the expected behavior when a message cannot be completed successfully?

Reply:
Yes. We should terminate the downstream. I want to wrap this with a retry().

Reply:
This operator is just to complete/abandon. . . if it fails, it'll pass it downstream. the retries and infinite one is out of scope for this PR.